src/librustdoc/html/render/span_map.rs

```diff
@@ -0,0 +1,215 @@
+use crate::clean;
+use crate::html::sources;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit::{self, NestedVisitorMap, Visitor};
+use rustc_hir::{ExprKind, GenericParam, GenericParamKind, HirId, Mod, Node};
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+
+use std::path::{Path, PathBuf};
+
+/// This enum allows us to store two different kinds of information:
+///
+/// In case the `span` definition comes from the same crate, we can simply get the `span` and use
+/// it as is.
+///
+/// Otherwise, we store the definition `DefId` and will generate a link to the documentation page
+/// instead of the source code directly.
+#[derive(Debug)]
+crate enum LinkFromSrc {
+    Local(clean::Span),
+    External(DefId),
+}
+
+/// This struct is used only as index in the `span_map`, not as [`Span`]! `Span`s contain
+/// some extra information (the syntax context) we don't need. **Do not convert this type back to
+/// `Span`!!!**
```

Comment:
Why not? Because information loss?

Reply:
I think the comment is pretty explicit already.

Reply:
@GuillaumeGomez I don't agree: it says not to convert it back, but not why. Is it because you've lost the SyntaxContext?

Reply:
No, it's really because they are used only as key for the "span map" and shouldn't be used for anything else. They have no other purpose.